server/src/data_sources/mod.rs

```diff
@@ -1,4 +1,28 @@
+use unleash_types::client_features::ClientFeature;
+
+use crate::types::EdgeToken;
+
 pub mod builder;
 pub mod memory_provider;
 pub mod offline_provider;
 pub mod redis_provider;
+
+trait ProjectFilter<T> {
+    fn filter_by_projects(&self, token: &EdgeToken) -> Vec<T>;
+}
+
+impl ProjectFilter<ClientFeature> for Vec<ClientFeature> {
+    fn filter_by_projects(&self, token: &EdgeToken) -> Vec<ClientFeature> {
+        self.iter()
+            .filter(|feature| {
+                if let Some(feature_project) = &feature.project {
+                    token.projects.contains(&"*".to_string())
+                        || token.projects.contains(feature_project)
+                } else {
+                    false
+                }
+            })
+            .cloned()
```

Comment:
Not a big deal, but this code is pretty similar to the one in server/src/data_sources/redis_provider.rs, maybe something to extract

Reply:
Yeah, basically the same code. It'll be the same code for every source. Nice target for a trait impl in future